File: JAVA/310. Minimum Height Trees.java
class Solution {
    public List<Integer> findMinHeightTrees(int n, int[][] edges) {
        List<Integer> result = new ArrayList<>();
        if (n == 1) {
            result.add(0);
            return result;
        }
        List<Set<Integer>> adj = new ArrayList<>(n);
        for (int i = 0; i < n; ++i)
            adj.add(new HashSet<>());
        for (int[] edge : edges) {
            adj.get(edge[0]).add(edge[1]);
            adj.get(edge[1]).add(edge[0]);
        }
        Queue<Integer> leaves = new LinkedList<>();
        for (int i = 0; i < n; ++i)
            if (adj.get(i).size() == 1)
                leaves.add(i);
        while (n > 2) {
            int size = leaves.size();
            n -= size;
            for (int i = 0; i < size; ++i) {
                int leaf = leaves.poll();
                int neighbor = adj.get(leaf).iterator().next();
                adj.get(neighbor).remove(leaf);
                if (adj.get(neighbor).size() == 1)
                    leaves.add(neighbor);
            }
        }
        while (!leaves.isEmpty())
            result.add(leaves.poll());
        return result;
    }
}

File: C++/310. Minimum Height Trees.cpp
class Solution {
public:
    vector<int> findMinHeightTrees(int n, vector<vector<int>>& edges) {
        if(n <= 2) {
            vector<int> result;
            result.push_back(0);
            if (n == 2) result.push_back(1);
            return result;
        }
        vector<vector<int>> graph (n);
        for(vector<int> & edge : edges) {
            graph[edge[0]].push_back(edge[1]);
            graph[edge[1]].push_back(edge[0]);     
        }
        vector<int> height(n);
        height[0] = findHeight(graph, 0, -1, height);
        vector<int>  result;
        int u = 0;
        while(result.size() == 0) {
            int max1vertex, max1 = -2;
            int max2vertex, max2 = -2;
            for(int v : graph[u]) {
                if(max1 == -2) {
                    max1 = height[v];
                    max1vertex = v;
                }
                else if(height[v] >= max1) {
                    max2 = max1;
                    max2vertex = max1vertex;
                    max1 = height[v];
                    max1vertex = v;
                }
                else if(max2 == -2 || height[v] > max2) {
                    max2 = height[v];
                    max2vertex = v;
                }
            }            
            if(max1 == max2) {      
                result.push_back(u);
            } 
            else if(max1 == max2 + 1) {    
                result.push_back(u);
                result.push_back(max1vertex);
            } 
            else {
                  height[u] = (max2 == -2) ? 0 : max2 + 1;
                  u = max1vertex; 
             }        
        }
        return result;
    }    
    private:
        int findHeight(vector<vector<int>>& graph, int u, int parent, vector<int>& height) {
            int maxHeight = -1;
             for(int v : graph[u]) {
                 if(v != parent) {
                     height[v] = findHeight(graph, v, u, height);
                     maxHeight = max(maxHeight, height[v]);
                 }
             }
            return maxHeight + 1;
        }
};
